smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpProtocolGeneratorUtils.java

```diff
@@ -286,4 +292,96 @@ static void writeHostPrefix(GenerationContext context, OperationShape operation)
             });
         });
     }
+
+    /**
+     * Writes a function used to dispatch event to corresponding event deserializers.
+     * This function assumes a event-specific deserialization function is generated
+     * for each returned structures.
+     *
+     * @param context The generation context.
+     * @param events The union of of events bond to an event stream
+     * @return A list of all event structure shapes for the event stream that were dispatched to.
+     */
+    static Set<StructureShape> generateDeserializingEventUnion(
+            GenerationContext context,
+            UnionShape events
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Symbol symbol = symbolProvider.toSymbol(events);
+        String protocolName = context.getProtocolName();
+        String methodName = ProtocolGenerator.getDeserFunctionName(symbol, protocolName) + "_event";
+        Model model = context.getModel();
+        Set<StructureShape> targets = new TreeSet<>();
+        writer.openBlock("const $L = async (\n"
+                + "  output: any,\n"
+                + "  context: __SerdeContext\n"
+                + "): Promise<$T> => {", "}", methodName, symbol, () -> {
+            events.getAllMembers().forEach((name, member) -> {
+                Shape target = model.expectShape(member.getTarget());
+                targets.add(target.asStructureShape().orElseThrow(
+                        () -> new CodegenException("Expect event to be structure shape, got " + target.getType())));
+                // In shape deser, Unions are indexed by local name. But event name in event stream has different
+                // meaning than actual unions index, which may have local name. Here assume event union is indexed by
+                // member name
+                writer.openBlock("if (output['$L'] !== undefined) {", "}", name, () -> {
+                    writer.openBlock("return {", "};", () -> {
+                        // Dispatch to special event deserialize function
+                        Symbol eventSymbol = symbolProvider.toSymbol(target);
+                        String eventDeserMethodName =
```

Comment:
This building process can be condensed quite a bit:

```java
Symbol eventSymbol = symbolProvider.toSymbol(target);
writer.write("$1L: await $2L(output[$1S], context)", name,
        ProtocolGenerator.getDeserFunctionName(eventSymbol, protocolName) + "_event");
```

This kind of clean up can be done in a couple places.

Reply:
I don't think this would work because `context` is not passed to each union member's deser function.